File: solutions/kotlin/luhn/1/src/main/kotlin/Luhn.kt
object Luhn {

    fun isValid(candidate: String): Boolean {
        // 1. Убираем все пробелы
        val cleanCandidate = candidate.replace(" ", "")

        // 2. Длина должна быть строго больше 1 символа
        if (cleanCandidate.length <= 1) {
            return false
        }

        var sum = 0
        // Флаг: нужно ли удваивать текущую цифру. 
        // Идем справа налево, поэтому вторую цифру справа (индекс с конца) удваиваем.
        var doubleCurrent = false

        // 3. Перебираем строку с конца в начало
        for (i in cleanCandidate.length - 1 downTo 0) {
            val char = cleanCandidate[i]

            // Если встретили не цифру — номер неверный
            if (!char.isDigit()) {
                return false
            }

            var digit = char.digitToInt()

            // Удваиваем каждую вторую цифру, считая справа
            if (doubleCurrent) {
                digit *= 2
                if (digit > 9) {
                    digit -= 9
                }
            }

            sum += digit
            // Меняем флаг для следующей цифры
            doubleCurrent = !doubleCurrent
        }

        // 4. Номер верен, если сумма делится на 10 без остатка
        return sum % 10 == 0
    }
}

File: solutions/julia/luhn/1/luhn.jl
function luhn(cardnum)
    # 1. Удаляем все пробелы из строки
    cleaned = replace(cardnum, " " => "")
    
    # 2. Строки длиной 1 или меньше невалидны
    if length(cleaned) <= 1
        return false
    end
    
    # 3. Проверяем, что строка состоит только из цифр
    if !all(isdigit, cleaned)
        return false
    end
    
    total_sum = 0
    # Идем с конца строки к началу
    # Переменная double флаг: удваивать ли текущую цифру (начинаем с false для последней цифры)
    double = false
    
    for i in length(cleaned):-1:1
        # Конвертируем символ в число
        digit = parse(Int, cleaned[i])
        
        if double
            digit *= 2
            if digit > 9
                digit -= 9
            end
        end
        
        total_sum += digit
        double = !double  # Меняем флаг для следующей цифры
    end
    
    # 4. Номер валиден, если сумма делится на 10 без остатка
    return total_sum % 10 == 0
end
